### PR TITLE
MLP Adjustments

### PR DESCRIPTION
**Pull request by JFranklin3000**
Closes [issue URL]

##### In this PR
- MLP Health Adjustments
- MLP Essence Select

##### Testing
- Validate drop is working for MLP Roles
- Build a role with type MLP

### CODE REVIEW

File: module/sheet-handlers/role-handler.mjs [hunk 1]
```diff
@@ -88,7 +78,142 @@ export class RoleHandler {
...
+
+    return isUnique;
+  }
+
+  /**
+   * Handles Setting the Values of what was selected in the Essence Selection Dialog
+   * @param {Object} options The selections made in the dialog window.
+   * @param {Object} role The role that was dropped on the actor
+   * @param {Function} dropFunc The function for the drop of the character
+   */
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:162]:
```suggestion
  /**
   * Handles setting the values of what was selected in the Essence Selection Dialog
   * @param {Object} options The selections made in the dialog window.
   * @param {Role} role The role that was dropped on the actor
   * @param {Function} dropFunc The function for the drop of the role
   */
```

File: module/sheet-handlers/role-handler.mjs [hunk 2]
```diff
@@ -88,7 +78,142 @@ export class RoleHandler {
...
+    if (newRole.system.adjustments.health.length) {
+      const totalIncrease = await roleValueChange(this._actor.system.level, newRole.system.adjustments.health);
+      const newHealthBonus = this._actor.system.health.bonus + totalIncrease;
+
+      await this._actor.update({
+        "system.health.bonus": newHealthBonus,
+      });
+    }
+
+    await createItemCopies(newRole.system.items, this._actor, "perk", newRole);
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:217]: This looks a lot like `onLevelChange()`, can we consolidate?

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: module/sheet-handlers/advancement-handler.mjs [hunk 3]
```diff
@@ -49,6 +49,15 @@ export class AdvancementHandler {
         });
       }
 
+      if (item.system.adjustments.health.length) {
+        const totalIncrease = await roleValueChange(this._actor.system.level, item.system.adjustments.health, previousLevel);
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheet-handlers/advancement-handler.mjs:53]: fixed, good call.

File: module/helpers/utils.mjs [hunk 4]
```diff
@@ -122,6 +122,21 @@ export function rememberOptions(html) {
   return options;
 }
 
+/**
+ * Returns values of inputs upon dialog submission. Used for passing data between sequential dialogs.
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/helpers/utils.mjs:126]: removed

File: module/helpers/utils.mjs [hunk 5]
```diff
@@ -122,6 +122,21 @@ export function rememberOptions(html) {
...
+ * @param {HTML} html   The html of the dialog upon submission
+ * @returns {Object>}  The dialog inputs and their submitted values
+ * @private
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/helpers/utils.mjs:129]: removed

File: module/helpers/utils.mjs [hunk 6]
```diff
@@ -122,6 +122,21 @@ export function rememberOptions(html) {
   return options;
 }
 
+/**
+ * Returns values of inputs upon dialog submission. Used for passing data between sequential dialogs.
+ * @param {HTML} html   The html of the dialog upon submission
+ * @returns {Object>}  The dialog inputs and their submitted values
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/helpers/utils.mjs:128]: fixed

File: module/sheet-handlers/role-handler.mjs [hunk 7]
```diff
@@ -29,37 +30,26 @@ export class RoleHandler {
...
+      try {
+        const essenceTypes = await this._essenceSelect(role,dropFunc);
+        if (!essenceTypes) {
+          console.warn("Essences Not selected");
+        } else {
+          console.log(essenceTypes);
+        }
+      } catch(error) {
+        console.error(error);
+      }
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheet-handlers/role-handler.mjs:45]: But now it is fine. so removed.

File: templates/dialog/essence-select.hbs [hunk 8]
```diff
@@ -0,0 +1,42 @@
+<form class="essence-select-dialog">
+  <div class="flexcol item-sheet">
+    <div class="essence-select">
+      <div>
+        {{#> "systems/essence20/templates/item/parts/sheet-field.hbs" label='E20.EssenceStrength'}}
+        {{#*inline "item-field-inputs"}}
+          <select class="item-field-inputs" name="strength" id="strength">
+            {{selectOptions choices nameAttr="key" labelAttr="label"}}
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply templates/dialog/essence-select.hbs:8]: Changed

File: templates/actor/parts/misc/pc-skills.hbs [hunk 9]
```diff
@@ -2,7 +2,11 @@
   {{!-- Strength --}}
   <div class="skills-container" style="border-color: {{system.color}};">
     <div class="essence-header strength">
-      <span class="essence-label">{{localize config.essences.strength}}</span>
+      {{#if system.ponyEssence.strength}}
+        <span class="essence-label {{system.ponyEssence.strength}}">{{localize config.essences.strength}}</span>
+      {{else}}
+        <span class="essence-label">{{localize config.essences.strength}}</span>
+      {{/if}}
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply templates/actor/parts/misc/pc-skills.hbs:9]: Sounds good.

File: module/sheet-handlers/role-handler.mjs (hunk 1 above)
Comment (JFranklin3000) [reply module/sheet-handlers/role-handler.mjs:162]: fixed

File: module/sheet-handlers/role-handler.mjs (hunk 2 above)
Comment (JFranklin3000) [reply module/sheet-handlers/role-handler.mjs:217]: I think we ran into this initially that it was just different enough that we couldn't quite put them together, but if we can figure it out I am all for it.

File: template.json [hunk 10]
```diff
@@ -109,6 +109,12 @@
             "total": 0
           }
         },
+        "ponyEssence": {
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply template.json:112]: Went with essenceRanks I think that is better.

File: lang/en.json [hunk 11]
```diff
@@ -215,6 +215,11 @@
     "EssenceBoostsSpeed": "Speed Essence Boosts",
     "EssenceBoostsStrength": "Strength Essence Boosts",
     "EssenceIncrease": "Select Essence Score Increase Option",
+    "EssenceProgressionBronze": "Bronze",
+    "EssenceProgressionDiamond": "Diamond",
+    "EssenceProgressionGold": "Gold",
+    "EssenceProgressionSilver": "Silver",
```
Comment (phildominguez) [inline lang/en.json:221]: `EssenceRank*`

File: lang/en.json [hunk 12]
```diff
@@ -215,6 +215,11 @@
...
+    "EssenceSelect": "Select Essence Progression",
```
Comment (phildominguez) [inline lang/en.json:222]: `EssenceProgressionSelect`

File: module/helpers/config.mjs [hunk 13]
```diff
@@ -815,3 +815,10 @@ E20.MLPAdvancement = {
   ],
 };
 
+E20.AdvancementNames = [
```
Comment (phildominguez) [inline module/helpers/config.mjs:818]: `EssenceRankNames`

File: module/sheet-handlers/role-handler.mjs [hunk 14]
```diff
@@ -88,7 +67,112 @@ export class RoleHandler {
...
     await deleteAttachmentsForItem(role, this._actor);
     this._actor.setFlag('essence20', 'previousLevel', 0);
   }
+
+  /**
+   * Handles the selection of the Essence Progression Ranks
+   * @param {Object} role The role that was dropped on the actor
+   * @param {Function} dropFunc The function for the drop of the character
+   */
+  async _essenceSelect(role, dropFunc) {
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:97]: `_selectEssenceProgression`

File: module/sheet-handlers/role-handler.mjs [hunk 15]
```diff
@@ -88,7 +67,112 @@ export class RoleHandler {
...
+      },
+    ).render(true);
+  }
+
+  /**
+   * Handles verifying that all of the Essences have a different rank
+   * @param {Object} options The selections made in the dialog window.
+   * @returns {Boolean} If all of the Essences have a different rank
+   */
+  _verifySelection (options) {
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:132]:
```suggestion
  _verifyEssenceProgression(options) {
```

File: module/sheet-handlers/role-handler.mjs [hunk 16]
```diff
@@ -88,7 +67,112 @@ export class RoleHandler {
...
+    return isUnique;
+  }
+
+  /**
+   * Handles setting the values of what was selected in the Essence Selection Dialog
+   * @param {Object} options The selections made in the dialog window.
+   * @param {Role} role The role that was dropped on the actor
+   * @param {Function} dropFunc The function for the drop of the role
+   */
+  async _essenceSetValues (options, role, dropFunc) {
```
Comment (phildominguez) [inline module/sheet-handlers/role-handler.mjs:152]:
```suggestion
  async _setEssenceProgression(options, role, dropFunc) {
```

**Review by phildominguez — CHANGES_REQUESTED** [review]

**Review by phildominguez — APPROVED** [review]

**Pull request merged**